### PR TITLE
GEODE-7778: Add PUBLISH, SUBSCRIBE and UNSUBSCRIBE Redis commands

### PR DESCRIPTION
**Pull request by sabbeyPivotal**
Redis PUBLISH/SUBSCRIBE is described here: https://redis.io/topics/pubsub

This implementation introduces 3 new command classes to support this:
`SubscribeExecutor`, `PublishExecutor` and `UnsubscribeExecutor`. At a high
level, the flow is as follows:
- A Redis client issues a SUBSCRIBE command. The server maintains a list of
  subscribed clients and associate channel(s) in the `Subscribers` class. A
  subscribed `Client` uses Netty's notion of a `uniqueID` to identify itself
  uniquely.
- When a client PUBLISHes a message on a channel, the publish 'event' uses an
  `onMembers` function call to distribute the channel/message to all members.
  Each member then publishes the message to any `Client`s registered on the
  channel.

Points to note:
Redis pub/sub is not HA. i.e. A subscriber only receives those messages
published after it has successfully subscribed. If the client is disconnected
and reconnects, it will not receive messages published during the disconnect
window.

When clients subscribe, an open connection is established to the server and the
client blocks waiting for messages.

This implementation does not require messages to be published on the same
server to which the clients, executing subscribe, are connected since the
published message is distributed to all servers.

Co-authored-by: Jens Deppe <jdeppe@pivotal.io>
Co-authored-by: Prasath Durairaj <vdurairaj@pivotal.io>
Co-authored-by: Murtuza Boxwala <mboxwala@pivotal.io>
Co-authored-by: Sarah Abbey <sabbey@pivotal.io>
Co-authored-by: John Hutchison <jhutchison@pivotal.io>
Co-authored-by: Ray Ingles <ringles@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    return totalSubscribers;
+  }
+
+  public long subscribe(String channel, ExecutionHandlerContext context, Client client) {
+    if (subscribers.exists(channel, client)) {
+      return subscribers.findSubscribers(client).size();
+    }
+    Subscriber subscriber = new Subscriber(client, channel, context);
+    subscribers.add(subscriber);
+    return subscribers.findSubscribers(client).size();
```
Comment (kohlmu-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: Are we concerned about thread safety here or not? Because, if we would have done a count of all subscribers before adding, then we don't have to iterate all subscribers again just to in all reality do `previousCount + 1`

File: geode-redis/src/main/java/org/apache/geode/redis/internal/ExecutionHandlerContext.java [hunk 2]
```diff
@@ -86,6 +88,11 @@ public KeyRegistrar getKeyRegistrar() {
   }
 
   private KeyRegistrar keyRegistrar;
+  private PubSub pubSub;
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/ExecutionHandlerContext.java]: Only one `PubSub` is created for the lifetime of the adapter. Each `ExecutionContextHandler` receives a reference to it.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.apache.geode.cache.execute.Function;
+import org.apache.geode.cache.execute.FunctionContext;
+import org.apache.geode.cache.execute.FunctionService;
+import org.apache.geode.cache.execute.ResultCollector;
+
+public class PubSub {
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: I think that would definitely be a good thing if we find there is a need for multiple implementations. For now it's just a concrete class.

[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: Can we then just add an interface and an implementation... It is easier to do this BEFORE rather than after. Also it makes it easier to achieve loose coupling.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  private void prune(List<Subscriber> failedSubscribers) {
+    failedSubscribers.forEach(subscriber -> {
+      if (subscriber.client.isDead()) {
+        subscribers.remove(subscriber.client);
+      }
+    });
+  }
+
+  private class Subscribers {
+    List<Subscriber> subscribers = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: We're coding to the interface and not the concrete class. It could easily be changed if the need arises.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/PublishExecutor.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import org.apache.geode.redis.internal.Command;
+import org.apache.geode.redis.internal.ExecutionHandlerContext;
+import org.apache.geode.redis.internal.Executor;
+
+public class PublishExecutor implements Executor {
+  @Override
+  public void executeCommand(Command command, ExecutionHandlerContext context) {
+    byte[] channelName = command.getProcessedCommand().get(1);
+    byte[] message = command.getProcessedCommand().get(2);
+    long publishCount = context.getPubSub().publish(new String(channelName), new String(message));
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/PublishExecutor.java]: It's a good point. We realize that `context` is very overloaded right now and we're planning on refactoring in the future.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/SubscribeExecutor.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+public class SubscribeExecutor extends AbstractExecutor {
+
+  @Override
+  public void executeCommand(Command command, ExecutionHandlerContext context) {
+    ArrayList<ArrayList<Object>> items = new ArrayList<>();
+    for (int i = 1; i < command.getProcessedCommand().size(); i++) {
+      ArrayList<Object> item = new ArrayList<>();
+      byte[] channelName = command.getProcessedCommand().get(i);
+      long subscribedChannels =
+          context.getPubSub().subscribe(new String(channelName), context, context.getClient());
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/SubscribeExecutor.java]: Ditto above.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/UnsubscribeExecutor.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.geode.redis.internal.ExecutionHandlerContext;
+import org.apache.geode.redis.internal.executor.AbstractExecutor;
+
+public class UnsubscribeExecutor extends AbstractExecutor {
+
+  @Override
+  public void executeCommand(Command command, ExecutionHandlerContext context) {
+    byte[] channelName = command.getProcessedCommand().get(1);
+    long subscriptionCount =
+        context.getPubSub().unsubscribe(new String(channelName), context.getClient());
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/UnsubscribeExecutor.java]: Same ditto

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: ArrayLists create heavier constructs under the hood, regardless of "I'm coding against that interface". All I'm asking is, given that you know you will never use  indexing, would a LinkedList not suffice?

File: geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+import java.util.Arrays;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+
+class Subscriber {
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: I don't think that will work since `Comparable` only uses a single type and `isEqualTo` requires multiple types.

[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: Yes thank you. I do see that now.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  public Subscriber(Client client, String channel,
+      ExecutionHandlerContext context) {
+
+    this.client = client;
+    this.channel = channel;
+    this.context = context;
+  }
+
+  public boolean isEqualTo(String channel, Client client) {
+    return channel.equals(this.channel) && client.equals(this.client);
```
Comment (kohlmu-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: There is no null checking happening. Can we really assume that the input `channel` and `client` are non-null?

File: geode-redis/src/main/java/org/apache/geode/redis/internal/RedisCommandType.java [hunk 10]
```diff
@@ -2595,6 +2598,69 @@ public RedisDataType getDataType() {
...
+    @Override
+    public RedisDataType getDataType() {
+      return this.dataType;
+    }
+  },
+  PUBLISH {
+    private Executor executor;
+
+    @Override
+    public Executor getExecutor() {
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/RedisCommandType.java]: Not really, but since this pattern is used throughout all the commands I think it would be better to make that change in a separate PR and cover all the commands.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/RedisCommandType.java [hunk 11]
```diff
@@ -2595,6 +2598,69 @@ public RedisDataType getDataType() {
...
+  /**
+   * SUBSCRIBE channel...
+   * <p>
+   * subscribe to channel
+   */
+  SUBSCRIBE {
+    private Executor executor;
+
+    @Override
+    public Executor getExecutor() {
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/RedisCommandType.java]: Ditto

File: geode-redis/src/main/java/org/apache/geode/redis/internal/RedisCommandType.java [hunk 12]
```diff
@@ -2595,6 +2598,69 @@ public RedisDataType getDataType() {
...
+    @Override
+    public RedisDataType getDataType() {
+      return this.dataType;
+    }
+  },
+  UNSUBSCRIBE {
+    private Executor executor;
+
+    @Override
+    public Executor getExecutor() {
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/RedisCommandType.java]: Ditto

File: geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  public boolean publishMessage(String channel, String message) {
+    ByteBuf messageByteBuffer;
+    try {
+      messageByteBuffer = Coder.getArrayResponse(context.getByteBufAllocator(),
+          Arrays.asList("message", channel, message));
+    } catch (CoderException e) {
+      throw new RuntimeException(e);
```
Comment (kohlmu-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: Would we not prefer a checked exception here?

File: geode-redis/src/main/java/org/apache/geode/redis/internal/RedisCommandType.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/RedisCommandType.java]: Could we please raise one and address accordingly

File: geode-redis/src/main/java/org/apache/geode/redis/internal/ExecutionHandlerContext.java [hunk 14]
```diff
@@ -86,6 +88,11 @@ public KeyRegistrar getKeyRegistrar() {
...
+
+  public PubSub getPubSub() {
+    return pubSub;
+  }
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/ExecutionHandlerContext.java]: It is used by the various Executors that deal with pub/sub. Once `ExecutionContextHandler` is refactored, this should not be necessary.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    FunctionService.registerFunction(new Function() {
+      @Override
+      public String getId() {
+        return REDIS_PUB_SUB_FUNCTION_ID;
+      }
+
+      @Override
+      public void execute(FunctionContext context) {
+        String[] publishMessage = (String[]) context.getArguments();
+        long subscriberCount = publishMessageToSubscribers(publishMessage[0], publishMessage[1]);
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: This function call is fairly tightly controlled - higher level calling methods should already ensure correct parameterization.

File: geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java [hunk 16]
```diff
@@ -730,4 +740,11 @@ private static String getLogLevel(String arg) {
     }
     return logLevel;
   }
+
+  private PubSub createPubSub() {
+    PubSub pubSub = new PubSub();
+    pubSub.registerPublishFunction();
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java]: It ensures that the `PubSub` instance is fully constructed before use - since the Function requires a reference to it.

[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java]: Let me rephrase, given that there are no dependencies as input parameters, the registering of Functions can happen at construction time.

File: geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java [hunk 17]
```diff
@@ -446,6 +449,7 @@ private void initializeRedis() {
         throw assErr;
       }
       this.keyRegistrar = new KeyRegistrar(redisMetaData);
+      this.pubSub = createPubSub();
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java]: That is true.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: Can I quote you on this? Can we please add validation BEFORE invoking.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+    for (long subscriberCount : subscriberCounts) {
+      totalSubscribers += subscriberCount;
+    }
+    return totalSubscribers;
+  }
+
+  public long subscribe(String channel, ExecutionHandlerContext context, Client client) {
+    if (subscribers.exists(channel, client)) {
+      return subscribers.findSubscribers(client).size();
+    }
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: I'm not sure I understand. We need to know two things:
1 - has a client already subscribed to a channel
2 - how many subscriptions a client currently has

In order to address both of these concerns, the `findSubscribers` method would probably need to be inlined which seems like it would be less readable.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java (hunk 1 above)
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: I think that's just a different approach - also valid, but just different....

File: geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java (hunk 9 above)
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: Null checking is handled by the respective `equals` methods.

Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: So, you can 100% guarantee that the fields are ALWAYS non-null?

Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: another approach is that the equals method is `this.channel.equals(channel`. This way the constructor can ensure that the instance channel field is non-null AND your `isEqualsTo` is null-safe

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: You are correct. 
That said.. What is the max number of Pub/Sub combinations we can expect? How does this approach scale?
Maybe two maps <channel,List<Subscriber>> and <client,List<Subscriber>> ?

File: geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+    CountDownLatch latch = new CountDownLatch(1);
+
+    ChannelFutureListener channelFutureListener = future -> latch.countDown();
+
+    ChannelFuture channelFuture =
+        context.writeToChannelWithListener(messageByteBuffer, channelFutureListener);
+
+    try {
+      latch.await(1, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: Fixed code

File: geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java (hunk 9 above)
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: Yes, you're correct.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java (hunk 13 above)
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: Removed this exception in favor of returning `null`

File: geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java]: But I don't see any immediate value in changing it.

[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java]: Also, you can use `new PubSub` as the `registerFunction` can (and should) be called in the constructor of `PubSub` otherwise one is exposing knowledge that does not need to be exposed.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: Added validation

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: Could we leave optimization until it's actually determined that it is needed?

File: geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java (hunk 13 above)
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: Removed the exception in favor of an explicit return

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: I agree, premature optimization.... But I'd be interested in what the "normal" usage on this would be. Maybe we add a load testing ticket on this, to see what load this construct can take. Is it 100's or 1000's or even 10,000's...

But I think it would be good to understand when it will slow down / break.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/PubSub.java]: Done

File: geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java]: Done

File: geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java (hunk 9 above)
Comment (kohlmu-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: This implementation is not "optimal". If you makes sure the `channel` and `client` on the `Subscriber` is non-null, then your `isEqualsTo` method implementation changes. Then the comparison `this.channel.equals(channel) && this.client.equals(client)` is cleaner. There is no requirement to continually check if the method parameters are non-null, because it does not matter.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/PublishExecutor.java [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+  public void executeCommand(Command command, ExecutionHandlerContext context) {
+    List<byte[]> args = command.getProcessedCommand();
+    if (args.size() != 3) {
+      command.setResponse(Coder.getErrorResponse(context.getByteBufAllocator(), ArityDef.PUBLISH));
+      return;
+    }
+
+    String channelName = new String(args.get(1));
+    String message = new String(args.get(2));
+    long publishCount = context.getPubSub().publish(channelName, message);
```
Comment (kohlmu-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/PublishExecutor.java]: Feature envy. Can we please change.

Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/PublishExecutor.java]: I've raised GEODE-7793 (Refactor ExecutionHandlerContext to avoid 'feature-envy') to address this.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java (hunk 9 above)
Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriber.java]: Fixed
